drivers/hmc5883l/hmc5883l.c

```diff
@@ -0,0 +1,338 @@
+
+    if (memcmp(id_r, id_c, HMC5883L_ID_LEN)) {
+        DEBUG_DEV("sensor is not available, wrong id %02x%02x%02x, "
+                  "should be %02x%02x%02x",
+                  dev, id_r[0], id_r[1], id_r[2], id_c[0], id_c[1], id_c[2]);
+        return -HMC5883L_ERROR_WRONG_ID;
+    }
+
+    return res;
+}
+
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit)
+{
+    ASSERT_PARAM(byte != NULL);
+
+    uint8_t shift = 0;
+    while (!((mask >> shift) & 0x01)) {
+        shift++;
+    }
+    *byte = ((*byte & ~mask) | ((bit << shift) & mask));
+}
+
+static uint8_t _get_reg_bit(uint8_t byte, uint8_t mask)
+{
+    uint8_t shift = 0;
+
+    while (!((mask >> shift) & 0x01)) {
+        shift++;
+    }
+    return (byte & mask) >> shift;
+}
+
+static int _reg_read(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(data != NULL);
+    ASSERT_PARAM(len != 0);
+
+    DEBUG_DEV("read %d byte from sensor registers starting at addr 0x%02x",
+              dev, len, reg);
+
+    if (i2c_acquire(dev->params.dev)) {
+        DEBUG_DEV("could not aquire I2C bus", dev);
+        return -HMC5883L_ERROR_I2C;
+    }
+
+    int res = i2c_read_regs(dev->params.dev, dev->params.addr, reg, data, len, 0);
+    i2c_release(dev->params.dev);
+
+    if (res == HMC5883L_OK) {
+        if (ENABLE_DEBUG) {
+            printf("[hmc5883l] %s i2c dev=%d addr=%02x: read following bytes: ",
+                   __func__, dev->params.dev, dev->params.addr);
+            for (int i = 0; i < len; i++) {
+                printf("%02x ", data[i]);
+            }
+            printf("\n");
+        }
+    }
+    else {
+        DEBUG_DEV("could not read %d bytes from sensor registers "
+                  "starting at addr %02x, reason %d (%s)",
+                  dev, len, reg, res, strerror(res * -1));
+        return -HMC5883L_ERROR_I2C;
+    }
+
+    return res;
+}
+
+static int _reg_write(const hmc5883l_t *dev, uint8_t reg, uint8_t *data, uint16_t len)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(data != NULL);
+    ASSERT_PARAM(len != 0);
+
+    DEBUG_DEV("write %d bytes to sensor registers starting at addr 0x%02x",
+              dev, len, reg);
+
+    if (ENABLE_DEBUG) {
+        printf("[hmc5883l] %s i2c dev=%d addr=%02x: write following bytes: ",
+               __func__, dev->params.dev, dev->params.addr);
+        for (int i = 0; i < len; i++) {
+            printf("%02x ", data[i]);
+        }
+        printf("\n");
+    }
+
+    if (i2c_acquire(dev->params.dev)) {
+        DEBUG_DEV("could not aquire I2C bus", dev);
+        return -HMC5883L_ERROR_I2C;
+    }
+
+    int res;
+
+    if (!data || !len) {
+        res = i2c_write_byte(dev->params.dev, dev->params.addr, reg, 0);
+    }
+    else {
+        res = i2c_write_regs(dev->params.dev, dev->params.addr, reg, data, len, 0);
+    }
+    i2c_release(dev->params.dev);
+
+    if (res != HMC5883L_OK) {
```

Comment:
should be `(res != 0)`, as `0` is the official `OK` value for the I2C driver...

Reply:
Good point.